internal/unionstore/art/node.go

```diff
@@ -0,0 +1,741 @@
+	// By reducing 4 bits for nodeAddr, we can save about 1/4-1/3 memories for node.
+	artNodeSize = 12
+	node4size   = 80
+	node16size  = 236
+	node48size  = 888
+	node256size = 3096
+	leafSize    = 12
+)
+
+var (
+	// nullNode48 is used to initialize the node 256
+	nullNode48 = node48{}
+	// nullNode256 is used to initialize the node 256
+	nullNode256 = node256{}
+)
+
+var nullArtNode = artNode{kind: typeInvalid, addr: nullAddr}
+
+func init() {
+	for i := 0; i < node48cap; i++ {
+		nullNode48.children[i] = artNode{kind: typeInvalid, addr: nullAddr}
+	}
+	for i := 0; i < node256cap; i++ {
+		nullNode256.children[i] = artNode{kind: typeInvalid, addr: nullAddr}
+	}
+}
+
+type Key []byte
+type artNode struct {
+	kind nodeKind
+	addr nodeAddr
+}
+
+type node struct {
+	nodeNum     uint8
+	prefixLen   uint8
+	prefix      [maxPrefixLen]byte
+	inplaceLeaf artNode
+}
+
+type node4 struct {
+	node
+	present  uint8 // bitmap, only lower 4-bit is used
+	keys     [node4cap]byte
+	children [node4cap]artNode
+}
+
+type node16 struct {
+	present uint16 // bitmap
+	node
+	keys     [node16cap]byte
+	children [node16cap]artNode
+}
+
+// Node with 48 children
+const (
+	n48s = 6  // 2^n48s == n48m
+	n48m = 64 // it should be sizeof(node48.present[0])
+)
+
+type node48 struct {
+	node
+	present  [4]uint64
+	keys     [node256cap]uint8 // map byte to index
+	children [node48cap]artNode
+}
+
+type node256 struct {
+	node
+	children [node256cap]artNode
+}
+
+type leaf struct {
+	vAddr nodeAddr
+	klen  uint16
+	flags uint16
+}
+
+func (n *artNode) isLeaf() bool {
+	return n.kind == typeLeaf
+}
+
+func (n *artNode) leaf(a *artAllocator) *leaf {
+	return a.getLeaf(n.addr)
+}
+
+func (n *artNode) node(a *artAllocator) *node {
+	switch n.kind {
+	case typeNode4:
+		return &n.node4(a).node
+	case typeNode16:
+		return &n.node16(a).node
+	case typeNode48:
+		return &n.node48(a).node
+	case typeNode256:
+		return &n.node256(a).node
+	default:
+		panic("invalid node kind")
+	}
+}
+
+func (n *artNode) at(a *artAllocator, idx int) artNode {
+	switch n.kind {
+	case typeNode4:
+		return n.node4(a).children[idx]
+	case typeNode16:
+		return n.node16(a).children[idx]
+	case typeNode48:
+		n48 := n.node48(a)
+		return n48.children[n48.keys[idx]]
+	case typeNode256:
+		return n.node256(a).children[idx]
+	default:
+		panic("invalid node kind")
+	}
+}
+
+func (n4 *node4) init() {
+	// initialize basic node
+	n4.nodeNum = 0
+	n4.prefixLen = 0
+	n4.inplaceLeaf = nullArtNode
+	// initialize node4
+	n4.present = 0
+}
+
+func (n4 *node4) nextPresentIdx(start int) int {
+	mask := math.MaxUint8 - (uint8(1) << start) + 1 // e.g. offset=3 => 0b11111000
+	present := n4.present & mask
+	present |= 0xf0
+	return bits.TrailingZeros8(present)
+}
+
+func (n4 *node4) prevPresentIdx(start int) int {
+	mask := uint8(1<<(start+1) - 1) // e.g. start=3 => 0b000...0001111
+	present := n4.present & mask
+	zeros := bits.LeadingZeros8(present)
+	return 8 - zeros - 1
+}
+
+func (n16 *node16) init() {
```

Comment:
Can we somehow reduce duplicated code in all `init`s?
Or can we consider using generics to reduce more duplicated code? Maybe a benchmark is needed.